src/audio/mfcc/mfcc_init.c

```diff
@@ -0,0 +1,453 @@
+		val = Q_MULTSR_32X32((int64_t)inv_cepstral_lifter, PI_Q23 * i, 21, 23, 23);
+		n = val / TWO_PI_Q23;
+		val -= n * TWO_PI_Q23;
+		sin = sin_fixed_32b(Q_SHIFT_LEFT(val, 23, 28)); /* Q4.28 -> Q1.31 */
+		/* Val is Q7.9 make 0.5 multiply with additional shift */
+		val = Q_MULTSR_32X32((int64_t)sin, cl->cepstral_lifter, 31, 9, 9 - 1);
+		val += ONE_Q9;
+		mat_set_scalar_16b(cl->matrix, 0, i, sat_int16(val));
+	}
+
+	return 0;
+}
+
+/* TODO mfcc setup needs to use the config blob, not hard coded parameters.
+ * Also this is a too long function. Split to STFT, Mel filter, etc. parts.
+ */
+int mfcc_setup(struct comp_dev *dev, int max_frames, int sample_rate, int channels)
+{
+	struct mfcc_comp_data *cd = comp_get_drvdata(dev);
+	struct sof_mfcc_config *config = cd->config;
+	struct mfcc_state *state = &cd->state;
+	struct mfcc_fft *fft = &state->fft;
+	struct psy_mel_filterbank *fb = &state->melfb;
+	struct mfcc_dct *dct = &state->dct;
+	int ret;
+
+#ifdef MFCC_DEBUGFILES
+	mfcc_init_debug_open();
+	mfcc_generic_debug_open();
+#endif
+
+	comp_dbg(dev, "mfcc_setup()");
+
+	/* Check currently hard-coded features to match configuration request */
+	if (!config->round_to_power_of_two || !config->snip_edges ||
+	    config->subtract_mean || config->use_energy) {
+		comp_err(dev, "mfcc_setup(): Can't change currently hard-coded features");
+		return -EINVAL;
+	}
+
+	if (config->sample_frequency != sample_rate) {
+		comp_err(dev, "mfcc_setup(): Config sample_frequency does not match stream");
+		return -EINVAL;
+	}
+
+	cd->max_frames = max_frames;
+	state->sample_rate = sample_rate;
+	state->low_freq = config->low_freq;
+	state->high_freq = (config->high_freq == 0) ? (sample_rate >> 1) : config->high_freq;
+	if (state->low_freq > state->high_freq) {
+		comp_err(dev, "mfcc_setup(): Config high_freq must be larger than low_freq");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "mfcc_setup(), source_channel = %d, stream_channels = %d",
+		  config->channel, channels);
+	if (config->channel >= channels) {
+		comp_err(dev, "mfcc_setup(): Illegal channel");
+		return -EINVAL;
+	}
+
+	if (config->channel < 0)
+		state->source_channel = 0;
+	else
+		state->source_channel = config->channel;
+
+	state->emph.enable = config->preemphasis_coefficient > 0;
+	state->emph.coef = -config->preemphasis_coefficient; /* Negate config parameter */
+	fft->fft_size = config->frame_length;
+	fft->fft_padded_size = 1 << (31 - norm_int32(fft->fft_size)); /* Round up to nearest 2^N */
+	fft->fft_hop_size = config->frame_shift;
+	fft->half_fft_size = (fft->fft_padded_size >> 1) + 1;
+
+	comp_info(dev, "mfcc_setup(), emphasis = %d", config->preemphasis_coefficient);
+	comp_info(dev, "mfcc_setup(), fft_size = %d", fft->fft_size);
+	comp_info(dev, "mfcc_setup(), fft_padded_size = %d", fft->fft_padded_size);
+	comp_info(dev, "mfcc_setup(), fft_hop_size = %d", fft->fft_hop_size);
+	comp_info(dev, "mfcc_setup(), fft->half_fft_size = %d", fft->half_fft_size);
```

Comment:
mmh, maybe we could somehow reduce the number of prints...

Reply:
Yes it can be compacted to one trace line. I think the config blob decode information is useful to see that a correct binary blob has been set up in topology.

Reply:
I deleted half_fft_size as redundant info and combined to same print line more stuff